homeassistant/components/picnic/coordinator.py

```diff
@@ -0,0 +1,149 @@
+"""Coordinator to fetch data from the Picnic API."""
+import copy
+from datetime import timedelta
+import logging
+
+import async_timeout
+from python_picnic_api import PicnicAPI
+from python_picnic_api.session import PicnicAuthError
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_ACCESS_TOKEN
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import ADDRESS, CART_DATA, LAST_ORDER_DATA, SLOT_DATA
+
+
+class PicnicUpdateCoordinator(DataUpdateCoordinator):
+    """The coordinator to fetch data from the Picnic API at a set interval."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        picnic_api_client: PicnicAPI,
+        config_entry: ConfigEntry,
+    ):
+        """Initialize the coordinator with the given Picnic API client."""
+        self.picnic_api_client = picnic_api_client
+        self.config_entry = config_entry
+        self._user_address = None
+
+        logger = logging.getLogger(__name__)
+        super().__init__(
+            hass,
+            logger,
+            name="Picnic coordinator",
+            update_interval=timedelta(minutes=30),
+        )
+
+    async def _async_update_data(self) -> dict:
+        """Fetch data from API endpoint."""
+        try:
+            # Note: asyncio.TimeoutError and aiohttp.ClientError are already
+            # handled by the data update coordinator.
+            async with async_timeout.timeout(10):
+                data = await self.hass.async_add_executor_job(self.fetch_data)
+
+            # Update the auth token in the config entry if applicable
+            await self._update_auth_token()
+
+            # Return the fetched data
+            return data
+        except ValueError as error:
+            raise UpdateFailed(f"API response was malformed: {error}") from error
+        except PicnicAuthError as error:
+            raise UpdateFailed(f"API authentication expired: {error}") from error
```

Comment:
We have a specific exception for this.

```suggestion
            raise ConfigEntryAuthFailed from error
```